pkg/identity/cache/allocator.go

```diff
@@ -235,7 +282,26 @@ func Release(ctx context.Context, id *identity.Identity) (bool, error) {
 		return false, fmt.Errorf("allocator not initialized")
 	}
 
-	return IdentityAllocator.Release(ctx, globalIdentity{id.Labels})
+	lastUse, err := IdentityAllocator.Release(ctx, globalIdentity{id.Labels})
+
+	if err != nil {
+		return false, err
+	}
+
+	idty := idpool.ID(id.ID.Uint32())
+	identityRefCountMutex.Lock()
+	if refCount := idPoolRefCount[idty]; refCount > 0 {
+		lastRef := refCount == 1
+		if lastRef {
+			// As it is the last reference for this identity we can safely remove
+			// its controller
+			identityControllerManager.RemoveControllerAndWait(fmt.Sprintf("sync-identity (%d)", idty))
```

Comment:
Not 100 % sure on this - couldn't this potentially block for a while if the controller is running and sleeping for 5 minutes within its `DoFunc`? Since we need to wait for the controller which is running to currently exit to  Is blocking for that long a desirable behavior?

Reply:
There's a context in the controller's `DoFunc`, it should be canceled when the controller is called.